### ISSUE TITLE
[C#] feature("director") + %shared_ptr(Type) resolves in uncompilable C# code  

### ISSUE THREAD

**Issue by OlegJakushkin** [swig/swig#419]
# Problem

in SWIG 3.5 use of `feature("director")` + `%shared_ptr(Type)` resolves  in uncompilable C# code generation.
# Example code

**.i**

```
%module(directors="1") M
%include "std_shared_ptr.i"

%feature("director"); // Manual shared_ptr into needed type conversion is required! Stuff like SWIGTYPE_p_std__shared_ptrT_%TYPE_NAME%_t into TYPE_NAME replacement
%shared_ptr(A)
%shared_ptr(B)

%include "SimplifiedAPI.h"
```

**.h**

```
class A {
    virtual std::string SayHello();
    virtual ~A();
};

class B {
public:
    virtual void SetA(std::shared_ptr<A> p);
    virtual ~B();
};
```

**Error:** `SWIGTYPE_p_std__shared_ptrT_A_t.cs`  gets generated and makes does not allow C# code to compile.
# How to get around:

On C# side it generates `SWIGTYPE_p_std__shared_ptrT_%TYPE_NAME%_t` constructor call that has to be manually converted into into `%TYPE_NAME%`  call.
 `SWIGTYPE_p_std__shared_ptrT_%TYPE_NAME%_t.cs` file can be removed manually

**Comment by wsfulton** [comment]
The private members in A above actually causes a segfault in SWIG, which I've fixed in 54e2317b24de05ad38e7670d76726d8797e12851. I suspect that A is posted incorrectly in this ticket and the intention was for the members to be public rather than private.

**Comment by OlegJakushkin** [comment]
Yes it was public - sory, my bad=)

**Comment by wsfulton** [comment]
I don't have time to debug and test this properly, however, the obvious missing typemap in boost_shared_ptr.i (which is used by std_shared_ptr.i for std::shared_ptr) is to add in the missing csdirectorin typemap. Try this:
```
%typemap(csdirectorin) std::shared_ptr<A> "new A($iminput, true)"
```

**Comment by Maxhy** [comment]
Same issue on my side with latest version (3.0.8) and I can indeed fix it according to @wsfulton advice, see bellow (where *logicalaccess::NdefMessage* is my class).
```
%typemap(csdirectorin) SWIG_SHARED_PTR_QNAMESPACE::shared_ptr< logicalaccess::NdefMessage > "new NdefMessage($iminput, true)"
%typemap(csdirectorout) SWIG_SHARED_PTR_QNAMESPACE::shared_ptr< NdefMessage > "NdefMessage.getCPtr($cscall).Handle"
```

But I would like to have something generic (I have hundreds of objects), if someone have an idea (I believe this should be considered as a bug btw)... Thanks.

**Comment by Deadpikle** [comment]
**SWIG 4.x**: It appears as though this isn't needed in SWIG 4.x. Is that accurate? In my project, I no longer needed the typemaps once I updated to SWIG 4.0.1 -- my project compiled cleanly. I don't even see the results of the `csdirectorout` showing up in my generated files (although maybe I am missing it and/or my code doesn't apply).

**SWIG 3.x**: We ran into a bug where the given `csdirectorin` samples were giving us valid C# objects that were wrappers around `null`, which broke things when we tried to use an object that was wrapping a null object. This is because the aforementioned `csdirectorin` typemaps don't have a null check. To fix this, you can adjust the typemaps to be like the following (the changes are within the " marks):

`%typemap(csdirectorin) SWIG_SHARED_PTR_QNAMESPACE::shared_ptr< CPPClassNameHere > "($iminput  == global::System.IntPtr.Zero) ? null  : new ClassNameHere($iminput, true)"`

**Comment by ojwb** [comment]
With git master, `SWIGTYPE_p_std__shared_ptrT_A_t.cs` is no longer generated.

Looks like this was indeed fixed in SWIG 4.0.0, by 46ea2ab0a6825fbb4600cbd6a821e710263016ed.
